pkg/server/helpers.go

```diff
@@ -205,34 +204,40 @@ func getRepoDigestAndTag(namedRef reference.Named, digest imagedigest.Digest, sc
 
 // localResolve resolves image reference locally and returns corresponding image metadata. It returns
 // nil without error if the reference doesn't exist.
+// Cases that the arg ref of localResolve is:
+// 1. regular imageID(e.g. sha256:9d7e6df8e5ca703a7629a6d47f9d6f90c3c4419203913d7b761248469bb0b364)
+// 2. imageName or repoTag or repoDigest:  we get imageID from runtime
+// 3. truncated imageID(e.g.  9d or sha256:9d)
 func (c *criContainerdService) localResolve(ctx context.Context, ref string) (*imagestore.Image, error) {
-	_, err := imagedigest.Parse(ref)
-	if err != nil {
-		// ref is not image id, try to resolve it locally.
-		normalized, err := util.NormalizeImageRef(ref)
-		if err != nil {
-			return nil, fmt.Errorf("invalid image reference %q: %v", ref, err)
-		}
-		image, err := c.client.GetImage(ctx, normalized.String())
-		if err != nil {
-			if errdefs.IsNotFound(err) {
-				return nil, nil
+	if _, err := imagedigest.Parse(ref); err != nil {
+		runtimeImageID, err := func(ref string) (string, error) {
+			// ref is not image id, try to resolve it locally.
+			normalized, err := util.NormalizeImageRef(ref)
+			if err != nil {
+				return "", fmt.Errorf("invalid image reference %q: %v", ref, err)
 			}
-			return nil, fmt.Errorf("failed to get image from containerd: %v", err)
-		}
-		desc, err := image.Config(ctx)
-		if err != nil {
-			return nil, fmt.Errorf("failed to get image config descriptor: %v", err)
+			image, err := c.client.GetImage(ctx, normalized.String())
+			if err != nil {
+				return "", fmt.Errorf("failed to get image from containerd: %v", err)
+			}
+			desc, err := image.Config(ctx)
+			if err != nil {
+				return "", fmt.Errorf("failed to get image config descriptor: %v", err)
+			}
+			return desc.Digest.String(), nil
+		}(ref)
+		if err == nil {
```

Comment:
Did you want:
```
if err != nil {
   return runtimeImageID, err 
} 
ref = runtimeImageID
```

Reply:
No.  If it is fail, We need to try it as truncindex.

Reply:
Then why generate and return all those errors above in the func if you are not going to use them? I presumed from reading func that you only wanted to continue in the nil case.

Reply:
Then why generate and return all those errors above in func if you are not going to use them? I presumed from reading func that you only wanted to continue in the error == nil case and intended to return the actual errors generated.

Reply:
Will do

Reply:
done